File: apps/workflow-engine/src/nodes/LLMChat.ts
import type { NodeDefinition } from '../schemas/workflow.js';
import type {
  ExecutionContext,
  NodeData,
  NodeExecutionResult,
} from '../engine/types.js';
import type { INodeTypeDescription } from '../engine/nodeSchema.js';
import { BaseNode } from './BaseNode.js';

/**
 * LLM Proxy API types
 */
interface LLMProxyRequest {
  system_prompt: string;
  user_prompt: string;
  temperature: number;
  max_output_tokens: number;
}

interface LLMProxyResponse {
  response: string;
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
  };
}

/**
 * LLM Chat Node - Simple single-call LLM integration via proxy service
 *
 * This node makes a single request to the LLM proxy and returns the response.
 * No tool calling, no conversation history - just prompt in, response out.
 */
export class LLMChatNode extends BaseNode {
  readonly type = 'LLMChat';
  readonly description = 'Make a simple LLM call via proxy service';

  static readonly nodeDescription: INodeTypeDescription = {
    name: 'LLMChat',
    displayName: 'LLM Chat',
    icon: 'fa:comment-dots',
    description: 'Make a simple LLM call via proxy service',
    group: ['ai'],
    inputs: [{ name: 'main', displayName: 'Input', type: 'main' }],
    outputs: [
      {
        name: 'main',
        displayName: 'Output',
        type: 'main',
        schema: {
          type: 'object',
          properties: {
            response: { type: 'string', description: 'LLM response text' },
            usage: {
              type: 'object',
              description: 'Token usage statistics',
              properties: {
                promptTokens: { type: 'number', description: 'Tokens in prompt' },
                completionTokens: { type: 'number', description: 'Tokens in completion' },
                totalTokens: { type: 'number', description: 'Total tokens used' },
              },
            },
          },
        },
      },
    ],

    properties: [
      {
        displayName: 'LLM Proxy URL',
        name: 'proxyUrl',
        type: 'string',
        default: 'http://localhost:8000/run_llm',
        required: true,
        description: 'URL of the LLM proxy service',
      },
      {
        displayName: 'System Prompt',
        name: 'systemPrompt',
        type: 'string',
        default: '',
        description:
          'Instructions for how the AI should behave. Supports expressions: {{ $json.context }}',
        typeOptions: { rows: 4 },
      },
      {
        displayName: 'User Prompt',
        name: 'userPrompt',
        type: 'string',
        default: '',
        required: true,
        description:
          'The message to send to the AI. Supports expressions: {{ $json.question }}',
        typeOptions: { rows: 6 },
      },
      {
        displayName: 'Temperature',
        name: 'temperature',
        type: 'number',
        default: 0.7,
        description:
          'Controls randomness (0 = deterministic, 1 = creative). Range: 0-1',
        typeOptions: { minValue: 0, maxValue: 1, step: 0.1 },
      },
      {
        displayName: 'Max Tokens',
        name: 'maxTokens',
        type: 'number',
        default: 1024,
        description: 'Maximum number of tokens in the response',
        typeOptions: { minValue: 1, maxValue: 8192 },
      },
    ],
  };

  async execute(
    _context: ExecutionContext,
    nodeDefinition: NodeDefinition,
    inputData: NodeData[]
  ): Promise<NodeExecutionResult> {
    const proxyUrl = this.getParameter<string>(
      nodeDefinition,
      'proxyUrl',
      'http://localhost:8000/run_llm'
    );
    const systemPrompt = this.getParameter<string>(
      nodeDefinition,
      'systemPrompt',
      ''
    );
    const userPrompt = this.getParameter<string>(nodeDefinition, 'userPrompt');
    const temperature = this.getParameter<number>(
      nodeDefinition,
      'temperature',
      0.7
    );
    const maxTokens = this.getParameter<number>(
      nodeDefinition,
      'maxTokens',
      1024
    );

    const results: NodeData[] = [];

    // Process each input item
    for (const item of inputData.length > 0 ? inputData : [{ json: {} }]) {
      const response = await this.callLLMProxy(proxyUrl, {
        systemPrompt,
        userPrompt,
        temperature,
        maxTokens,
      });

      results.push({
        json: {
          ...item.json,
          response: response.text,
          usage: response.usage,
        },
      });
    }

    return this.output(results);
  }

  private async callLLMProxy(
    proxyUrl: string,
    options: {
      systemPrompt: string;
      userPrompt: string;
      temperature: number;
      maxTokens: number;
    }
  ): Promise<{ text: string; usage: Record<string, number> }> {
    const { systemPrompt, userPrompt, temperature, maxTokens } = options;

    const requestBody: LLMProxyRequest = {
      system_prompt: systemPrompt,
      user_prompt: userPrompt,
      temperature,
      max_output_tokens: maxTokens,
    };

    const response = await fetch(proxyUrl, {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`LLM Proxy error (${response.status}): ${error}`);
    }

    const data = (await response.json()) as LLMProxyResponse;

    if (!data.response) {
      throw new Error('No response from LLM Proxy');
    }

    return {
      text: data.response,
      usage: {
        promptTokens: data.usage?.prompt_tokens || 0,
        completionTokens: data.usage?.completion_tokens || 0,
        totalTokens: data.usage?.total_tokens || 0,
      },
    };
  }
}

File: apps/workflow-engine/src/nodes/llm_chat.py
"""LLM Chat node - single-turn LLM integration."""

from __future__ import annotations

from typing import Any, TYPE_CHECKING

from .base import (
    BaseNode,
    NodeTypeDescription,
    NodeInputDefinition,
    NodeOutputDefinition,
    NodeProperty,
    NodePropertyOption,
)

if TYPE_CHECKING:
    from ..engine.types import ExecutionContext, NodeData, NodeDefinition, NodeExecutionResult


class LLMChatNode(BaseNode):
    """LLM Chat node - single-turn LLM conversation."""

    node_description = NodeTypeDescription(
        name="LLMChat",
        display_name="LLM Chat",
        description="Send a message to an LLM and get a response",
        icon="fa:robot",
        group=["ai"],
        inputs=[NodeInputDefinition(name="main", display_name="Input")],
        outputs=[
            NodeOutputDefinition(
                name="main",
                display_name="Response",
                schema={
                    "type": "object",
                    "properties": {
                        "response": {"type": "string", "description": "LLM response text"},
                        "model": {"type": "string", "description": "Model used"},
                        "usage": {"type": "object", "description": "Token usage stats"},
                    },
                },
            )
        ],
        properties=[
            NodeProperty(
                display_name="Model",
                name="model",
                type="options",
                default="gemini-2.0-flash",
                options=[
                    NodePropertyOption(name="Mock (Testing)", value="mock"),
                    NodePropertyOption(name="Gemini 2.0 Flash", value="gemini-2.0-flash"),
                    NodePropertyOption(name="Gemini 1.5 Flash", value="gemini-1.5-flash"),
                    NodePropertyOption(name="Gemini 1.5 Pro", value="gemini-1.5-pro"),
                    NodePropertyOption(name="GPT-4o", value="gpt-4o"),
                    NodePropertyOption(name="GPT-4o Mini", value="gpt-4o-mini"),
                    NodePropertyOption(name="Claude Sonnet", value="claude-sonnet-4-20250514"),
                ],
            ),
            NodeProperty(
                display_name="System Prompt",
                name="systemPrompt",
                type="string",
                default="You are a helpful assistant.",
                description="System message to set assistant behavior",
                type_options={"rows": 3},
            ),
            NodeProperty(
                display_name="User Message",
                name="userMessage",
                type="string",
                default="{{ $json.message }}",
                required=True,
                description="Message to send. Supports expressions: {{ $json.message }}",
                type_options={"rows": 5},
            ),
            NodeProperty(
                display_name="Temperature",
                name="temperature",
                type="number",
                default=0.7,
                description="Controls randomness (0-1)",
            ),
            NodeProperty(
                display_name="Max Tokens",
                name="maxTokens",
                type="number",
                default=1024,
                description="Maximum response length",
            ),
        ],
    )

    @property
    def type(self) -> str:
        return "LLMChat"

    @property
    def description(self) -> str:
        return "Send a message to an LLM and get a response"

    async def execute(
        self,
        context: ExecutionContext,
        node_definition: NodeDefinition,
        input_data: list[NodeData],
    ) -> NodeExecutionResult:
        from ..engine.types import NodeData
        from ..engine.expression_engine import ExpressionEngine, expression_engine

        model = self.get_parameter(node_definition, "model", "mock")
        system_prompt_template = self.get_parameter(node_definition, "systemPrompt", "You are a helpful assistant.")
        user_message_template = self.get_parameter(node_definition, "userMessage", "{{ $json.message }}")
        temperature = self.get_parameter(node_definition, "temperature", 0.7)
        max_tokens = self.get_parameter(node_definition, "maxTokens", 1024)

        results: list[NodeData] = []

        for idx, item in enumerate(input_data if input_data else [NodeData(json={})]):
            # Resolve expression against current item's data
            expr_context = ExpressionEngine.create_context(
                input_data,
                context.node_states,
                context.execution_id,
                idx,
            )
            # Resolve expressions in both system prompt and user message
            system_prompt = expression_engine.resolve(system_prompt_template, expr_context)
            user_message = expression_engine.resolve(user_message_template, expr_context)

            if not user_message:
                raise ValueError("User message is required")

            # Use mock response for testing
            if model == "mock":
                result = self._mock_response(user_message)
            else:
                result = await self._call_llm(
                    model=model,
                    system_prompt=system_prompt,
                    user_message=user_message,
                    temperature=temperature,
                    max_tokens=max_tokens,
                )
            results.append(NodeData(json=result))

        return self.output(results)

    def _mock_response(self, user_message: str) -> dict[str, Any]:
        """Generate a mock response for testing."""
        lower_msg = user_message.lower().strip()

        # Number-based responses for testing dynamic behavior
        number_responses = {
            "1": "You entered 1! This is response ONE. 🥇",
            "2": "You entered 2! This is response TWO. 🥈",
            "3": "You entered 3! This is response THREE. 🥉",
            "4": "You entered 4! This is response FOUR. 🍀",
            "5": "You entered 5! This is response FIVE. ✋",
        }

        # Check for number input first
        if lower_msg in number_responses:
            response = number_responses[lower_msg]
            return {
                "response": response,
                "message": response,
                "model": "mock",
                "input": user_message,
                "usage": {"input_tokens": len(user_message.split()), "output_tokens": len(response.split())},
            }

        # Keyword-based responses
        keyword_responses = {
            "hello": "Hello! I'm a mock LLM assistant. Try entering numbers 1-5 for different responses!",
            "hi": "Hi there! I'm running in mock mode. Try: 1, 2, 3, 4, or 5 for dynamic responses!",
            "help": "Mock assistant here! Try these:\n• Enter 1-5 for numbered responses\n• 'hello' or 'hi' for greetings\n• 'joke' for a joke\n• 'weather' for weather\n• Anything else to see echo",
            "weather": "Mock weather report: Sunny with 72°F (22°C). Perfect coding weather! ☀️",
            "joke": "Why do programmers prefer dark mode? Because light attracts bugs! 🐛",
            "test": "Test successful! The workflow is working correctly. 🎉",
        }

        # Check for keyword matches
        for keyword, response in keyword_responses.items():
            if keyword in lower_msg:
                return {
                    "response": response,
                    "message": response,
                    "model": "mock",
                    "input": user_message,
                    "usage": {"input_tokens": len(user_message.split()), "output_tokens": len(response.split())},
                }

        # Echo response for anything else
        response = f"Echo: \"{user_message}\"\n\n(Mock mode - enter 1-5 or 'help' for more options)"
        return {
            "response": response,
            "message": response,
            "model": "mock",
            "input": user_message,
            "usage": {"input_tokens": len(user_message.split()), "output_tokens": len(response.split())},
        }

    async def _call_llm(
        self,
        model: str,
        system_prompt: str,
        user_message: str,
        temperature: float,
        max_tokens: int,
    ) -> dict[str, Any]:
        """Call LLM via the unified provider."""
        from ..engine.llm_provider import get_llm_provider, LLMMessage

        provider = get_llm_provider()

        messages: list[LLMMessage] = []
        if system_prompt:
            messages.append(LLMMessage(role="system", content=system_prompt))
        messages.append(LLMMessage(role="user", content=user_message))

        response = await provider.chat(
            model=model,
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
        )

        return {
            "response": response.content or "",
            "model": model,
            "usage": {
                "input_tokens": response.usage.input_tokens,
                "output_tokens": response.usage.output_tokens,
            },
        }
